File: Leetcode/Count Square Submatrices with All Ones/Count Square Submatrices with All Ones.py
# Problem Link : https://leetcode.com/problems/count-square-submatrices-with-all-ones/

# The intuition is to create a m*n array , namely 'dp' and at each cell, store the maximum area of the square
# with all 1 that could be formed and end at that cell, i.e. this cell is the bottom-right corner of the square.

# Starting from top-left corner of the dp, we store the number of such squares that can be formed.

# After each iteration we add the squares formed at that cell to variable 'ans' and return ans.

def count_squares(matrix):

    # dp to store the maximum no. of squares possible at that place
    dp = [[0 for i in range(len(matrix[0]))] for j in range(len(matrix))]

    # ans stores the total no. of squares formed.
    ans = 0

    # Looping through each cell of the given matrix.
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):

            # Corner cases as only square of dimensions [1*1] can be formed at 0th row and 0th column.
            if i == 0 or j == 0:
                dp[i][j] = matrix[i][j]

            else:
                if matrix[i][j] == 0:
                    dp[i][j] = 0
                else:

                    # Taking minimum of the left surrounding and adding one (for [1*1] square formed).
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1

            # updating our answer.
            ans += dp[i][j]
    return ans

# Example :

matrix = [
    [0, 1, 1, 1],
    [1, 1, 1, 1],
    [0, 1, 1, 1]
]
print(count_squares(matrix))

File: Leetcode/Count Square Submatrices with All Ones/Count Square Submatrices with All Ones.java
class Solution {
    public int countSquares(int[][] matrix) {
        //Answer will contain the maximum order that we can get
        int ans=0;
        //This will be our final answer
        int count=0;
        
        //Looping through the matrix and updating matrix value 
        // where the value at a position indicates that the number is the order
        //of square submatrix that can be found there
        for(int i=1;i<matrix.length;i++){
            for(int j=1;j<matrix[0].length;j++){
                if(matrix[i][j]!=0){
                    matrix[i][j]= Math.min(matrix[i][j-1], Math.min(matrix[i-1][j-1], matrix[i-1][j])) +1;
                    ans= Math.max(ans, matrix[i][j]);
                }            
            }
        }
        //This array will contain the number of order 0 matrices, number of order of 1 matrices and so on
        int[] matrices= new int[ans+1];
         for(int i=0;i<matrix.length;i++){
            for(int j=0;j<matrix[0].length;j++){
                //This is because if there is a 2x2 matrix, then it will contain 1x1 matrix also. Similarly if there is 3x3 matrix, then it will have 2x2 matrix and 1x1 matrix also
                matrices[matrix[i][j]]+= matrix[i][j];             
            }
        }
        //Finding total submatrices
        for(int i=0;i<matrices.length;i++){
            count+= matrices[i];
        }
        return count;
    }
}
